src/schematools/validation.py

```diff
@@ -79,192 +57,166 @@ def __init__(self, message: str) -> None:  # noqa: D107
         self.message = message
 
 
-class Validator:
-    """Base class for validators.
+_all: List[tuple[str, Callable[[DatasetSchema], Iterator[str]]]] = []
 
-    Not only is this a base class for validators,
-    it is also used for running all registered validators.
-    See Also: :meth:`run_all`
 
-    Registration is a side-effect of overriding this base class.
+def run_all(dataset: DatasetSchema) -> Iterator[ValidationError]:
+    r"""Run all registered validators.
 
-    """
-
-    _registry: ClassVar[List[Type[Validator]]] = []
-    dataset: DatasetSchema
+    Yields:
+        :class:`ValidationError`\s, if any.
 
-    @classmethod
-    def __init_subclass__(cls: Type[Validator]) -> None:
-        """Register sub classes."""
-        super().__init_subclass__()
-        cls._registry.append(cls)
+    """  # noqa: W605
+    for name, validator in _all:
+        for msg in validator(dataset):
+            yield ValidationError(validator_name=name, message=msg)
 
-    def __init__(self, dataset: DatasetSchema) -> None:
-        """Initialize the validator with a dataset.
 
-        Args:
-            dataset: The dataset to run the validations on.
-        """
-        self.dataset = dataset
+def _validator(name: str) -> Callable:
```

Comment:
perhaps call it `_register_validator()`? That's a bit more descriptive of what it does

Reply:
Done.